File: kh-sim/backends/cpp/tests/validation_test.cpp
// validation_test.cpp — KH-005 acceptance test
//
// Validates C++ diagnostics against kh_reference_output.json.
// Tolerance: +-5% on kinetic_energy and enstrophy (KH-PHYSICS.md S7).
// Incompressibility: divergence_rms < 1e-10.
//
// Usage: kh-sim-cpp-validate <path/to/kh_reference_output.json>
// CTest: cmake --build build && ctest --test-dir build -V

#include "physics.hpp"

#include <nlohmann/json.hpp>

#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>

using json = nlohmann::json;

static double pct_err(double got, double ref) {
    return std::abs((got - ref) / ref) * 100.0;
}

int main(int argc, char* argv[]) {
    // Path to reference JSON passed as first argument (or default relative path)
    std::string ref_path = "../../shared/physics/kh_reference_output.json";
    if (argc >= 2) ref_path = argv[1];

    // Load reference
    std::ifstream f(ref_path);
    if (!f) {
        std::cerr << "Cannot open reference file: " << ref_path << "\n"
                  << "Run: python kh-sim/shared/physics/kh_physics.py\n";
        return 1;
    }
    json ref_json = json::parse(f);

    int ref_nx    = ref_json.value("grid_nx", 64);
    int ref_ny    = ref_json.value("grid_ny", 32);
    int ref_steps = ref_json.value("steps_completed", 100);

    double ref_ke   = ref_json["diagnostics"]["kinetic_energy"];
    double ref_ens  = ref_json["diagnostics"]["enstrophy"];
    double ref_mv   = ref_json["diagnostics"]["max_vorticity"];
    double ref_div  = ref_json["diagnostics"]["divergence_rms"];

    // Build matching request
    khsim::SimulationRequest req;
    req.grid_nx                = ref_nx;
    req.grid_ny                = ref_ny;
    req.domain_lx              = 1.0;
    req.domain_ly              = 0.5;
    req.dt                     = 0.001;
    req.steps                  = ref_steps;
    req.reynolds_number        = 1000.0;
    req.velocity_shear         = 1.0;
    req.perturbation_amplitude = 0.01;
    req.perturbation_mode      = 2;

    auto result = khsim::simulate(req);
    const auto& d = result.diagnostics;

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "--- C++ vs Python reference ---\n";
    std::cout << "kinetic_energy : got=" << d.kinetic_energy
              << "  ref=" << ref_ke
              << "  err=" << std::setprecision(2) << pct_err(d.kinetic_energy, ref_ke) << "%\n";
    std::cout << "enstrophy      : got=" << std::setprecision(6) << d.enstrophy
              << "  ref=" << ref_ens
              << "  err=" << std::setprecision(2) << pct_err(d.enstrophy, ref_ens) << "%\n";
    std::cout << "max_vorticity  : got=" << std::setprecision(6) << d.max_vorticity
              << "  ref=" << ref_mv
              << "  err=" << std::setprecision(2) << pct_err(d.max_vorticity, ref_mv) << "%\n";
    std::cout << "divergence_rms : got=" << std::scientific << d.divergence_rms
              << "  ref=" << ref_div << "\n";

    // ── Acceptance criteria ────────────────────────────────────────────────────
    bool pass = true;
    auto check = [&](const char* name, double err_pct, double tol) {
        if (err_pct >= tol) {
            std::cerr << "FAIL " << name << ": " << err_pct << "% >= " << tol << "%\n";
            pass = false;
        }
    };
    check("kinetic_energy", pct_err(d.kinetic_energy, ref_ke), 5.0);
    check("enstrophy",      pct_err(d.enstrophy,      ref_ens), 5.0);
    if (d.divergence_rms >= 1e-10) {
        std::cerr << "FAIL divergence_rms: " << d.divergence_rms << " >= 1e-10\n";
        pass = false;
    }

    // ── Field shape check (nx*ny) ──────────────────────────────────────────────
    int expected_n = ref_nx * ref_ny;
    if ((int)result.u_velocity.size() != expected_n ||
        (int)result.vorticity.size()  != expected_n) {
        std::cerr << "FAIL field shape mismatch\n";
        pass = false;
    }

    if (pass) {
        std::cout << "\nAll tests PASSED\n";
        return 0;
    } else {
        std::cout << "\nSome tests FAILED\n";
        return 1;
    }
}

File: kh-sim/backends/rust/tests/validation_test.rs
/// validation_test.rs — KH-003 acceptance criteria
///
/// Validates Rust diagnostics against Python/NumPy canonical reference output.
/// Tolerance: ±5% on kinetic_energy and enstrophy (per KH-PHYSICS.md §7).
/// Incompressibility: divergence_rms < 1e-10.
///
/// Reference file: kh-sim/shared/physics/kh_reference_output.json
/// Run: cargo test --test validation_test
use std::path::PathBuf;

use serde::Deserialize;

// ── Mirror of reference JSON diagnostics block ────────────────────────────────

#[derive(Deserialize)]
struct RefDiagnostics {
    kinetic_energy:  f64,
    enstrophy:       f64,
    max_vorticity:   f64,
    divergence_rms:  f64,
}

#[derive(Deserialize)]
struct RefOutput {
    diagnostics: RefDiagnostics,
    grid_nx:     usize,
    grid_ny:     usize,
    steps_completed: usize,
}

// ── Load reference JSON ───────────────────────────────────────────────────────

fn load_reference() -> RefOutput {
    // Path relative to Cargo workspace root (where `cargo test` is run from)
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    // Traverse up to kh-sim/shared/physics/
    path.push("../../shared/physics/kh_reference_output.json");

    let raw = std::fs::read_to_string(&path).unwrap_or_else(|_| {
        panic!(
            "Cannot open reference file at {:?}. \
             Run `python kh_physics.py` in kh-sim/shared/physics/ first.",
            path
        )
    });
    serde_json::from_str(&raw).expect("Failed to deserialise reference JSON")
}

// ── Helpers ───────────────────────────────────────────────────────────────────

fn pct_err(got: f64, reference: f64) -> f64 {
    ((got - reference) / reference).abs() * 100.0
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[test]
fn test_diagnostics_within_tolerance() {
    let reference = load_reference();
    let ref_d = &reference.diagnostics;

    // Build a matching SimulationRequest (same grid / steps as reference)
    // We access the physics module directly via the binary crate.
    // Because this is an integration test (tests/ dir), we call the public API.
    // The binary crate name is "kh-sim-rust" — Rust normalises '-' to '_' for
    // the crate name used in use statements.
    use kh_sim_rust::physics::simulate;
    use kh_sim_rust::models::SimulationRequest;

    let req = SimulationRequest {
        grid_nx:               reference.grid_nx,
        grid_ny:               reference.grid_ny,
        domain_lx:             1.0,
        domain_ly:             0.5,
        dt:                    0.001,
        steps:                 reference.steps_completed,
        reynolds_number:       1000.0,
        velocity_shear:        1.0,
        perturbation_amplitude:0.01,
        perturbation_mode:     2,
        initial_omega:         None,
    };

    let result = simulate(&req);
    let got = &result.diagnostics;

    println!("--- Rust vs Python reference ---");
    println!("kinetic_energy : got={:.6}  ref={:.6}  err={:.2}%",
             got.kinetic_energy, ref_d.kinetic_energy,
             pct_err(got.kinetic_energy, ref_d.kinetic_energy));
    println!("enstrophy      : got={:.6}  ref={:.6}  err={:.2}%",
             got.enstrophy, ref_d.enstrophy,
             pct_err(got.enstrophy, ref_d.enstrophy));
    println!("max_vorticity  : got={:.6}  ref={:.6}  err={:.2}%",
             got.max_vorticity, ref_d.max_vorticity,
             pct_err(got.max_vorticity, ref_d.max_vorticity));
    println!("divergence_rms : got={:.2e}  ref={:.2e}",
             got.divergence_rms, ref_d.divergence_rms);

    // Acceptance criteria (KH-PHYSICS.md §7)
    assert!(
        pct_err(got.kinetic_energy, ref_d.kinetic_energy) < 5.0,
        "kinetic_energy error {:.2}% exceeds 5% tolerance",
        pct_err(got.kinetic_energy, ref_d.kinetic_energy)
    );
    assert!(
        pct_err(got.enstrophy, ref_d.enstrophy) < 5.0,
        "enstrophy error {:.2}% exceeds 5% tolerance",
        pct_err(got.enstrophy, ref_d.enstrophy)
    );
    assert!(
        got.divergence_rms < 1e-10,
        "divergence_rms {:.2e} exceeds incompressibility threshold 1e-10",
        got.divergence_rms
    );
}

#[test]
fn test_field_shapes() {
    use kh_sim_rust::physics::simulate;
    use kh_sim_rust::models::SimulationRequest;

    let req = SimulationRequest {
        grid_nx: 32, grid_ny: 16,
        domain_lx: 1.0, domain_ly: 0.5,
        dt: 0.001, steps: 10,
        reynolds_number: 1000.0, velocity_shear: 1.0,
        perturbation_amplitude: 0.01, perturbation_mode: 2,
        initial_omega: None,
    };

    let result = simulate(&req);
    let n = req.grid_nx * req.grid_ny;
    assert_eq!(result.u_velocity.len(), n, "u_velocity shape mismatch");
    assert_eq!(result.v_velocity.len(), n, "v_velocity shape mismatch");
    assert_eq!(result.vorticity.len(),  n, "vorticity shape mismatch");
    assert_eq!(result.pressure.len(),   n, "pressure shape mismatch");
    assert!((result.t_final - 0.01).abs() < 1e-9, "t_final mismatch");
}
